### PR TITLE
Merge HeliAttack into FlyAttack

### PR DESCRIPTION
**Pull request by reaperrr**
Split from #15803.

Depends on #16369.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Maybe this should queue `HeliFly` as a `ChildActivity` instead of doing all the positioning stuff itself?

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Maybe this complete branch should queue `HeliFly` as a `ChildActivity` instead of doing all the positioning stuff itself?

**Comment by reaperrr** [comment]
Waiting for #16483 to be merged first would likely save me from some serious rebase conflicts due to how HeliAttack currently handles movement, so I see little point in rebasing & updating this until #16483 is in.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: The problem with that, at least in this particular activity, is that it needs to check every tick whether it's close enough to the target now, so using a ChildActivity for movement would complicate things.

I agree that this activity shouldn't call SetPosition and TickFacing directly, though, and should at least use `Fly.FlyToward` from #16483 instead.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: But `HeliFly` already takes care of that, doesn't it? Why can't we just call `move.MoveWithinRange`, just like `AttackFollow` does?

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Thanks for pointing me in the right direction, adressed in #16483.

**Comment by reaperrr** [comment]
Updated and rebased.

Note: In light of the experience (or should I say trauma...?) with the HeliFly merge, *please* lets stick to merging and - apart from the 2nd commit, which is limited in scope and therefore relatively easy to handle - fixing regressions, if there are any.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: What is this code trying to do? Remember that facing 0 and 1 are real facings, -1 is used to mean "any facing".

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: If !CanHover we queue the takeoff and the activities to do after takeoff in one go, whereas CanHover queus just the takeoff and waits until this next ticks to queue the movement.

We should be able to reduce duplication by putting
```csharp
if (!isAirborne)
	QueueChild(self, new TakeOff(self), true);
```

before the `if (!aircraft.Info.CanHover)` and letting both cases queue their movements after it?

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: It would improve readability if we format the code as
```csharp
if (!aircraft.Info.CanHover)
{
	if (...)
		...
	...
}
else
{
	if (...)
		...
	...
}
```
instead of having the different CanHover cases handled as `else if`s. This comment can probably then be removed as being obvious.

**Comment by reaperrr** [comment]
Updated. I had the new commit almost ready anyway (though originally for a follow-up...).

If this regresses anything major/hard-to-fix, I'd rather drop it and defer those changes back to that follow-up again, though.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Was replacing `checkTarget` with `target` intentional? This will regress targeting vs units that have disappeared under the fog (`target` refers to the *actual* hidden unit, `checkTarget` to its last known position).

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Yes, it actually was intentional.

If `useLastVisibleTarget` is `true`, we won't get here anyway (due to the `if` block above), and if it's false, `checkTarget` is equal to `target` anyway.
This was just about consistency (this and some places below were using checkTarget, while other parts below already used target).
I can change everything to `checkTarget` if you think that's better, but like I said, if we got this far in the activity they're equal anyway.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Ah, now I see. What you have here is fine.

**Review by pchote — DISMISSED** [review]
LGTM, but IMO it would be be better if the commits were squashed before merging.

**Comment by reaperrr** [comment]
> but IMO it would be be better if the commits were squashed before merging.

Done.

**Review by tovl — APPROVED** [review]
LGTM :+1:

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5f6f9b5db6d20a0b12a691d24eddb3bb5550bda5

**Pull request merged**
